### ISSUE TITLE
Let's move PSModulePath out of the documents folder

### ISSUE THREAD

**Issue by JohnLudlow** [PowerShell/PowerShell#15552]
## Summary of the new feature/enhancement

The documents folder has always been where modules user-scoped are kept, and that's always been ok, but I recently discovered some problems with this as I've been getting more familiar with Azure and my work laptop maps Documents to OneDrive. This can be problematic for a few reasons, such as OneDrive not syncing correctly or trying to restore a bunch of files from a module I just uninstalled. This is particularly an issue with larger modules (or groups of modules) such as Az.*

We can certainly wish OneDrive was better at its job, but really these files shouldn't be there - they are not documents. The user is not expected to edit them or look at them. They are more akin to application files or application data. There's no value to them being in OneDrive other than to transfer them to another machine, and that can be better achieved by having a script in OneDrive that calls PSDepend or PowerShellGet to install modules on that other machine.

I have updated my profile script to remove this path but it reappears after running `install-module`.

## Proposed technical implementation details (optional)

1. Remove `Documents\PowerShell\Modules` from the `$env:PSModulePath` default. Select a new default path such as `~\PowerShell\Modules` or `$env:localappdata\PowerShell\Modules`
2. Offer a cmdlet or documentation to move the existing modules to the new location, and optionally update the value of the environment variable.
3. Additionally, offer to move the existing modules and update the value of the environment variable during the install. (That way the user has the choice to make the install take care of it or do it later at a time of their choosing).

**Comment by jeroenlandheer** [comment]
A few extra things to consider when putting this up for triage:

- When I install a module on one machine and I am later at a different PC, OneDrive updates the modules folder... but `Upgrade-Moduile` doesn't work anymore, because it says the module wasn't installed with `Install-Module` ... (which it was, just not at the machine I'm working on.) 
- If your module folder is big loading PowerShell can take a (very) long time because it will wait until OneDrive has pulled all the files to your machine.

I think this change should happen... hopefully soon. I think storage for these modules should be in the `%userpfile%\AppData\Local` folder, this is the place for things that should not get synced with other machines.

PS: I create PowerShell modules... I have a lot of them... and they are all on my OneDrive now. 😒

**Comment by iSazonov** [comment]
@jeroenlandheer Do you mean _Update-Module_? If so you couls discuss in PowerShellGet repository.

**Comment by JohnLudlow** [comment]
@iSazonov Would the PowerShellGet guys be able to change the default value of `$env:PSModulePath` and update the installer to optionally update it on upgrade?

Though there should probably be a second issue in PowerShellGet to respect the value, not overwrite it.

**Comment by iSazonov** [comment]
@JohnLudlow Any changes related to PSModulePath are breaking changes. We need to keep backward compatibility.

**Comment by jeroenlandheer** [comment]
@iSazonov Sorry, that was a typo. This should indeed go there too, but don't you think they will choose the location of the modules where they install them... based on whatever this teams says it should be?  (AFAIK there's no way to customize this with settings, but that's something for another day.)

If paths for _loading_ the modules is the primary function of `PSModulePath`, this issue can be solved quite easily:
1. Designate a folder somewhere in `%LocalAppData%\...` 
2. Tell others that installing modules in this folder is best practice.
3. Put _both_ paths in `PSModulePath` (i.e. `%LocalAppData%\..`  and `%userprofile%\Documents\...`)
4. Ensure the new location is added to new installations in the PSModulePath environment variable.

No need to move stuff around, new things get installed in `%LocalAppData%\..` and old modules that are still in the user's documents folder keep working. Maybe this doesn't even require a code change in PowerShell, just some changes in the installer.

And now we're on this subject, in Windows installations, modules which are installed on the local machine are installed by default in the `Program files\PowerShell\Modules` folder. I think the correct place for this should be `%AllUsersProfile%\...` (often called `C:\ProgramData`) 

AFAIK the functionality of `PSModulePath` itself doesn't change with all of this, so this is technically not a breaking change.

Hope this helps.

**Comment by JohnLudlow** [comment]
@jeroenlandheer Indeed, and if I do want to move the older modules to the new folder (I probably would, in my case), there are some options:

* Offer to do that in the installer
* Offer a tool (such as a cmdlet) to do it post-install
* Offer documentation that says "you can copy everything from `<old location>` to `<new location>`"

Any of those would work for me, but some people may appreciate it being done for them.

And you're absolutely correct IMHO about the Windows Modules. The only modules that should be there are ones that are installed as part of PowerShell and shouldn't be removed.

**Comment by SeeminglyScience** [comment]
> AFAIK the functionality of `PSModulePath` itself doesn't change with all of this, so this is technically not a breaking change.

While I agree that it would be hard to claim this is explicitly a breaking change of an API, it would still be disruptive.  There has never been a supported way of determining the module paths of different scopes, e.g. no `PSModuleInfo.AllUsersModuleRepository` or similar.  Any tools looking to interact with the module directories explicitly has been required to hard code the path, so any location change here will break these tools.

(Note that I'm not necessarily advocating against it, but it's something that needs to be considered)

**Comment by rjmholt** [comment]
Related: https://github.com/PowerShell/PowerShell/issues/7082#issuecomment-397710807

**Comment by kilasuit** [comment]
Previously discussed in #8069 too

**Comment by rjmholt** [comment]
So we reviewed this on the @PowerShell/powershell-committee today, but also this is something I've personally done some thinking on in the past.

The mangling on particular locations on the PSModulePath is something we've seen users impacted by repeatedly. For example: https://github.com/PowerShell/vscode-powershell/issues/2824#issuecomment-667355371.

Another important point is that the CurrentUser and AllUsers module paths are actually configurable in the powershell.config.json:

https://github.com/PowerShell/PowerShell/blob/e2c23fc5bd035251bed71413ad7c467c99dafd76/src/System.Management.Automation/engine/Modules/ModuleIntrinsics.cs#L1304-L1305

So there is a concept of both the *default* CurrentUser module path and the actual one.

However, there's currently no way to communicate the actual one, so the default is the implicit contract between PowerShell's module system (`Import-Module`/`Get-Module`/command lookup) and PowerShellGet. I was hoping to improve that in #7082, but even without that the contact being implicit means there's room to get it wrong.

Today it seems that both PowerShell 7 and PowerShellGet agree that the default CurrentUser module path lies under the `MyDocuments` path, but evidence suggests that this wasn't always the case.

The big issue now is that OneDrive configurations override the MyDocuments folder and cause this unexpected migration.

But additionally, the actual CurrentUsers module path could be overridden with a configuration. And this is something that things like the PowerShell Azure Functions worker would want to do given the radically different environment in which it operates. Except that there's no way for PowerShellGet to pick this up.

In terms of changing the default PSModulePath, there have been a lot of proposals to fix PSModulePath in various ways. Changing the CurrentUser path away from MyDocuments is going to break PowerShellGet, since it depends on the same location being used. However, the MyDocuments path clearly presents problems, and both the proposed alternatives (`$env:LOCALAPPDATA\PowerShell\Modules` and `$HOME\PowerShell\Modules`) make sense.

So in terms of recommendations:

- The Engine Working Group has also been asked to review this, to go into a bit more technical depth.
- An API of some kind that gives both the default and actual module path components is probably a good direction to go in.
- In terms of backward compatibility, tooling can look for this API and fall back to the traditional defaults.
- The committee didn't decide on whether or not the default path should be changed. We want a chance for the Engine WG to discuss this, and also want to talk to the OneDrive team to understand things a bit better here.

**Comment by JohnLudlow** [comment]
@rjmholt That's a great summary. Thanks.

We'll see what the Engine Working Group say

**Comment by JamesWTruher** [comment]
With behavior of OneDrive the WG agrees that we need to come up with a proper solution, however the scenarios have a high level of complexity which is exacerbated by the need that other PS tools also need to be able to use whatever solution is made available. 

The WG proposes that PS should provide an API which allows tools to query for the default location of modules.  The tool chain also needs to be able to query PS so the values are consistent.

We propose that an RFC is needed to define the behaviors, not only for PowerShell proper but for tools like PowerShellGet.

We think at least the following requirements need to be addressed in the RFC:
- where is the default location for PSModulePath (system and/or user)
- how may this default location be modified
- how is the API used by PS and the other tools
- how are misconfigurations handled (non-existing paths, multiple paths, etc)

**Comment by ShadowXVII** [comment]
+1. Nuget doesn't have this problem because it uses `Profile Root\.nuget` -- though perhaps long term the %LOCALAPPDATA% would make more sense for the user modules for PowerShell.

It's like syncing a .git repo in OneDrive! Unnecessary.

**Comment by JohnLudlow** [comment]
@mikhey Changing `PSModulePath` works until you do `Install-Module` which breaks it again

A `~/.pswh` folder (or whatever it gets called) would work, just as long as it's not somewhere OneDrive will try and mess with it

**Comment by rjmholt** [comment]
> Would a dot folder work here possibly?

As @rkeithhill said, those dot folders pollute `$HOME`, and even though they're not displayed, they still need to be processed by programs that examine the Home directory, creating a sort of tragedy-of-the-commons problem where more and more applications putting their configurations and data in a directory based in the Home directory adds up to slower filesystem performance for unrelated tasks.

PowerShell actually already conforms to the [XDG base directory specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) on *nix, which is a standard intended to solve this problem. PowerShell data files, including modules, are stored in `$env:XDG_DATA_HOME/powershell`, which defaults to `$HOME/.local/share/powershell`.

The issue then is that Windows doesn't really offer a clear answer here, because its philosophy is to separate users from configurers/application installers.

Worth noting is that the XDG standard is configurable with a sensible default, and I imagine we want the same, we just need to weigh configurability against complexity (mainly for anyone developing against our decision) and performance.

**Comment by dafzor** [comment]
@rkeithhill and dot files are supposed to also be hidden, but **Windows is not linux** so dot files are left always visible.

So I'd like to add my voice here in asking to please respect the platform conventions on Windows and use the appropriate AppData folder the same way Linux is using XDG.

As for accessibility for profile editing, I think that's already covered by using `<editor of choice> $PROFILE` which will work regardless of platform.

At most adding a `$PROFILEHOME` as an alias for `(Get-ChildItem $PROFILE).Directory.FullName` if the user wants to open the folder for any reason?

**Comment by codaamok** [comment]
I'm just stopping by to express my frustration with how things are today. I'm happy to read about positive discussions for change, though!

Trying to update `PSModulePath` in your profile or via registry ([like the docs suggest](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_psmodulepath?view=powershell-7.2#modifying-psmodulepath)) is pointless, because as previously mentioned, as soon as you use `Install-Module`, the env var is clobbered again with the undesired OneDrive Documents folder. You'll also notice that `Install-Module` seems to ignore `PSModulePath`, too, as it wants to always search in `MyDocuments\[Windows]PowerShell`:

![image](https://user-images.githubusercontent.com/6683266/147689452-08d1df2c-f4cb-4e71-9b90-68532f2c4ff4.png)

To workaround this, you need to update your MyDocuments (aka 'Personal') folder to no longer be in the OneDrive folder. [This seems to work well for me](https://support.microsoft.com/en-us/topic/configuration-of-the-my-documents-folder-dfd9a90d-8f80-18d6-e7cc-f1566fc3b10b):

1. Right-click My Documents (on the desktop), and then click Properties.
2. In the Target box, type the new path to the My Documents folder, or click Move to browse to the folder and if needed, create a new My Documents folder. If the path you type does not exist, you are prompted to confirm that you want to create a new folder.
3. Click ~Yes to move files from the old My Documents location to the new location, or~ click No if you do not want to move these files. Note that if you click No, the files in the old location are not deleted, but they are no longer visible from My Documents after the change. 

Figured I'd share this in case anyone needed it /shrug

**Comment by billti** [comment]
Just to plus one on the pain and urgency here. I just lost half a day trying to figure out exactly the issue @jeroenlandheer mentioned back on June last year, namely trying to Update-Module on a machine fails because I installed the module on a difference machine, and both are sync'ing my documents via OneDrive.

Beside the issue above this causes, I'm now syncing over 900MB of PowerShell files to my OneDrive account, and I only have two modules installed! (The 'Az' and the 'Microsoft.Graph' modules - which I suspect are in common usage).

I'm probably just going to stop using OneDrive or install for AllUsers, but it was time consuming and frustrating to figure this out.

**Comment by iSazonov** [comment]
I want to bring this problem to the attention of @PowerShell-Committee and @Area-PowerShellGet group.
I think it is good time to make the breaking change, preserve Document folder for module searching but install in new and more appropriate folder in user profile. The some for all user folder (machine-wide) - c:\ProgramData as right place. Also it makes sense to review where we place user profile (and, specially, profiles and config files of  host applications).
Maybe create RFC?
/cc @SteveL-MSFT

**Comment by ALIENQuake** [comment]
The `C:\ProgramData` was recently dropped by Git for Windows team due to the amount of confusion it creates: https://github.com/git-for-windows/git/pull/2358 Let's forget about this directory once for all.

**Comment by billti** [comment]
There are a number of teams at Microsoft that I'm sure have a ton of research and failures to learn from in this space. Nuget, npm, VS Code extensions, etc. seem to have solved similar problems on Windows in a cross-platform ecosystem. Folks like @zkat have a wealth of experience they might be able to share (or point us to people who can) to come to the best solution for PowerShell.

**Comment by RobertKlohr** [comment]
I want to add another reason to make this change.

If you have [Controlled Folder Access](https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/enable-controlled-folders) enabled then PowerShell is no longer able to install or update modules that reside in the default modules folder in the current user scope e.g., C:\Users\[user]\Documents\PowerShell or C:\Users\[user]\Documents\WindowsPowerShell.

The only way to fix this issue is to enable PowerShell for Controlled Folder Access bypass but doing that is explicitly against Microsoft's configuration recommend, and for good reason.

**Comment by JakeMoe** [comment]
I'm hitting this issue with Controlled Folder Access as well.  Our IT Apps department wants to start migrating things from legacy systems to newer PowerShell scripts, but they can't install the modules needed because Defender gets in the way, and they don't have admin access on their systems (only IT Operations have that).  I'd rather not blanket exclude powershell.exe / pwsh.exe from controlled folder access, as that sounds like a Very Bad Idea (tm).  I suppose I could deploy scripts via Intune that would install modules in AllUser scope, but that'd then require me to manage their module updates as well, and ideally the user should be able to manage their own modules.

**Comment by ShadowXVII** [comment]
Let's make a decision here! Snooze!

The Windows standard (AppData\Local) is fine. And makes sense. While I like dot paths, they're a *nix convention so can understand why the PowerShell team wouldn't apply this UNLESS they're not wanting a Windows-specific code change. This (I assume) is the same reason .dotnet and other teams have used dot paths.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee has re-discussed this recently.  The ideal solution is:

1. expose a public API in SMA.dll for apps to discover the system, shared, and user module paths.  Even our own modules like PowerShellGet are currently hardcoded to Documents (on Windows anyways)
2. ideally, $env:PSModulePath is JUST an env var and pwsh.exe doesn't manipulate it trying to infer user intentions (this was inherited behavior form powershell.exe), however, this would be a breaking change.  Basically, if certain system, shared, and user paths are not in the env var, PowerShell assumes this was a mistake and adds them except for some specific conditions
3. the actual location would probably be AppData on Windows.  On non-Windows, it's already outside of Documents.  Do we need a migration story for users with PS7 already installed?  A clean install is easy, but an existing PS7 means modules are likely already in Documents.  Seems to me a cmdlet to explicitly migrate may be best solution to avoid any surprises to the user.

Open to ideas/thoughts on this.

**Comment by iSazonov** [comment]
1. Since we say about fundamental behavior for years I think we need to put this in RFC.
2. We have no need to change system module path
3. There is a problem in how PowerShell process x86 and non-x86 paths for shared module paths. We should make this consistent. I think we have no need a migration path for the case (only if a experimental feature)
4. This change may be most sensitive for user module paths. Migration options:
    4.1 Experimental feature
    4.2 Installer warning
    4.3 Compatibility mode with chaining new path then old path.

**Comment by chrisbues** [comment]
Agree with the rest of the commenters that AppData makes the most sense on Windows. This also clears things up with module location path as well, given that OneDrive Known Folder Move's Documents redirection only applies at the explorer level.

**Comment by Jaykul** [comment]
Given that PowerShell already uses `$env:XDG_DATA_HOME/powershell` which defaults to `$HOME/.local/share/powershell` on Linux  ...

### Can we just let PowerShell on Windows support `$env:XDG_DATA_HOME` too?

PowerShell could just silently fall back to the current location if it's not set.

Then **users** could choose to set `$env:XDG_DATA_HOME` to `$Home` **or** `$env:LOCALAPPDATA` ... or `$env:APPDATA` 😨 ... or even the current default `$Home\Documents` or `$Home\OneDrive - MyCompany\Documents` 😆 

### There's no reason this should be a breaking change

Frankly, it can be done purely in PowerShellGet, because only users who set `$env:XDG_DATA_HOME` would need to also add `$env:XDG_DATA_HOME\.local\share\powershell\modules` to their `$env:PSModulePath`

**Comment by IISResetMe** [comment]
Engine WG discussed this issue today

## Points raised by the committee

We fully agree that SMA.dll should offer better insights into module path preferences through its public API surface - it's currently too opaque, and .

We also generally agree that PowerShell should, long-term, move towards `$env:LOCALAPPDATA` as the _preferred_ root for user-specific application data (modules included) on Windows, but certainly **not** as a breaking change. 
Instead, we recommend a "slow march" approach - establish and introduce a new path root, switch official host application logic defaults gradually, and continue supporting both paths for module discovery in parallel. 

Given the scope and expected impact of introducing a change like this (however carefully), the proposal should indeed be published as an RFC, and evaluated in the larger context of the ecosystem around PowerShell (tooling with hardcoded module paths, etc.). We don't have capacity to commit to producing the initial draft, but welcome additional input from the community and the committee.


## `XDG_DATA_HOME`

We further discussed @Jaykul's suggestion to support `$env:XDG_DATA_HOME` for establishing a user-defined path root for PowerShell-specific user data. 

We agree on its potential usefulness, but recommend **not** implementing support for `$env:XDG_DATA_HOME` in isolation. Instead, we would like to evaluate the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) holistically, and then give a more comprehensive recommendation on whether (and how) to support it consistently on Windows - please create a separate issue for this if that sounds valuable @Jaykul, we're happy to re-review it :)

**Comment by jeroenlandheer** [comment]
If I read all of this it seems that everyone is in agreement that the "My Documents" folder was a poor choice and that on Windows the AppData folder is preferred... but I notice one other thing.

Why does this have to take _more than a year_ to make a decision? Software engineers and administrators deal with breaking changes often enough, if something breaks we can handle it... it's not the end of the world. Anoyances like these are causing us more headaches (and time.)

I look forward to a preview/beta/rc to get this is resolved... soon? 😉

**Comment by haris1981** [comment]
One question, what will be the best practice if I want to use a single path for all the module installations (both Core and Windows Powershell) . As an organization we have multiple agent servers configured in DevOps tool (Octopus) and we have installed a set of PS commands in all the servers, so that Developers can use it while they deploy their code. And all the code deployment are running using Octopus service account which is admin in above mentioned Agent servers, so this will open the developers to install the new PS modules as and when needed and that may loosen the control over the installation path of PS modules. So what will be the best method if we want to control the module path (at least make it  2 paths- one for pwsh and another for windows ps) ?

**Comment by iSazonov** [comment]
@MHA1981 Please open question in Discussions.

**Comment by MHA1981** [comment]
Ok sure

**Comment by SteveL-MSFT** [comment]
This is a big enough change that it won't make it into 7.3 which is headed towards RC and we would not want to take such a impactful change at that time.  We should try to get something in early for 7.4 previews.

**Comment by bwilsonms** [comment]
do it! Otherwise I'll need to hack around on my box just so I don't have to wait for stuff to download from ODB

**Comment by JohnLudlow** [comment]
@SteveL-MSFT Is there a timeframe for 7.4 previews?

**Comment by SteveL-MSFT** [comment]
@JohnLudlow the first previews of 7.4 will be coming soon, but this feature won't show up til later

**Comment by fsackur** [comment]
I'd like to share the workaround we're implementing: junctions. OneDrive does not follow them, but they are transparent to most use cases.

IIUC, if you do `New-Item -ItemType Junction ~\Documents\PowerShell\Modules -Value ~\SomeOtherFolder\Modules`, you're good.

We are junctioning individual module folders within `Modules`, and that breaks PowerShellGet's update functionality in PSv5 due to `Get-ChildItem -Recurse` in PackageManagement\Get-Package, because `gci` doesn't traverse junctions within the provided path with `-Recurse`. (PSv7 does.) To work around that, we copy the `PSGetModuleInfo.xml` files back. PackageManagement then finds the XML files and uses the `InstalledLocation` property, which is junction-agnostic.

We didn't use symbolic links because they require admin to create on Windows.

**Comment by JustinGrote** [comment]
This is how I do it today for 7+:
1. Change my powershell.config.json PSModulePath to point to `%LocalAppData%/powershell/Modules`. This is the path on Linux/MacOS and puts it all in the same place everywhere.
2. Delete/migrate the modules folder from both "documents" folders for WinPS and PS and symlink (if I have admin) or ntfs junction (if I don't) to the localappdata folder.

This preserves functionality for PSGetv2/PSGetv3 to install modules. ModuleFast defaults to the localappdata location so it's not an issue there.

A low impact way to improve this situation I have proposed here: https://github.com/PowerShell/PowerShell/issues/18613 that would not change the default but at least make it configurable and exposed in a way that package managers can tie in. Once this is done both PSGetv3 and ModuleFast could detect this and utilize it (no hope for PSGetv2 since it is on life support tho)

**Comment by fowl2** [comment]
Having modules roam between machines is really nice functionality - it reduces friction switching between machines a lot.

Now I understand the existing mechanism is causing problems for some - luckily it's always worked for me.
1. Has the OneDrive team could be contacted to work out if there's changes that could be made on either side to make it more reliable?
1. I couldn't find any issue for a replacement roaming experience. eg. some sort of placeholder file, etc. although not all modules are available from the gallery, so quite a bit of design required there. Would this be welcome

On changing the path, please engage caution when breaking existing workflows by disabling/altering the currently enabled-by-default feature. 

(requested repost from https://github.com/PowerShell/PowerShellGet/issues/627#issuecomment-1322555942)

**Comment by SeeminglyScience** [comment]
> 1. Has the OneDrive team could be contacted to work out if there's changes that could be made on either side to make it more reliable?

Depends on the specific issues you were having. The biggest thing I'd recommend if you're trying to sync using a cloud provider is to ensure that your modules directory is considered "Pinned". Different providers call it different things, but for OneDrive you right click the folder and click this:

![image](https://user-images.githubusercontent.com/24977523/203396608-a50378b3-30f6-4038-9ce8-4ed87ec1af20.png)

> 2. I couldn't find any issue for a replacement roaming experience. eg. some sort of placeholder file, etc. although not all modules are available from the gallery, so quite a bit of design required there. Would this be welcome?

One potential option is to symlink the old location back to a folder in OneDrive

**Comment by zeratax** [comment]
is there already a way to move the `Microsoft.PowerShell_profile.ps1` file out of documents too?

I managed to get most out of my documents via the junctions suggestion, but I can't seem to move the profile file so I still end up with a PowerShell directory in my directory.

**Comment by o-l-a-v** [comment]
Can we include the scripts folder in this discussion too? Current behavior `<modulespath>\..\Scripts` seems like a hardcoded quick and dirty fix, an afterthought.

I think it should be treated separatly from the modules folder, at least have the ability to.

Maybe it could have it's own environmental variable, at least optionally?

* `PSScriptPath`?
* Default to `PSModulePath\..\Scripts` if `PSScriptPath` isn't set?

**Comment by MatejKafka** [comment]
Related: https://github.com/PowerShell/PowerShell/issues/13815

I support adding a runtime API to retrieve the preferred configuration directory, instead of just changing the default value to `$env:LOCALAPPDATA`. This would bring PowerShell one step closer to having configurable data/module/cache directories and encapsulated portable installs.

**Comment by thehappycheese** [comment]
I am here because I installed a big module pack (Azure PowerShell) and it crashed powershell because the modules directory is in an ultra slow UNC path network location. All morning to figure out what caused the slowdown. Now hacking PSModulePath to point elsewhere, but I will probably regret fiddling with it.

**Comment by o-l-a-v** [comment]
@thehappycheese 

Been there. Tip from one that have wasted hours on the same thing: PowerShellGet and PackageManagement does not deal with `PSModulePath`; path given scope is hard coded. But "Import-Module" works with `PSModulePath`.

So use [PackageManagement `Save-Package`](https://learn.microsoft.com/en-us/powershell/module/packagemanagement/save-package?view=powershell-7.3) with `-Path` to override where to install modules to.

**Comment by johnwmnolan** [comment]
@o-l-a-v  and @thehappycheese

Just remember that, in doing that, the modules are not technically "installed", only "saved", so you can't use Update-Module or Uninstall-Module to work with them. That's the (only?) drawback of using Save-Module over Install-Module.

**Comment by o-l-a-v** [comment]
@johnwmnolan 

Yup, good you pointed that out. I handle that with my own logic outside PowerShellGet and PackageManagement.

**Comment by JustinGrote** [comment]
> @o-l-a-v and @thehappycheese
> 
> Just remember that, in doing that, the modules are not technically "installed", only "saved", so you can't use Update-Module or Uninstall-Module to work with them. That's the (only?) drawback of using Save-Module over Install-Module.

The difference here is just an XML manifest file that gets dropped in the folder that's basically a CLIXML output. They're pretty easy to "fake" and make a module "installed" in PSGetv2 terms. PSGetv3 doesn't use this anymore.

**Comment by ShadowXVII** [comment]
Just an FYI -- Symlinks and Junctions (at least on Windows) **do** sync via OneDrive, so using the symlink workaround doesn't work.

**Comment by 4lex11s** [comment]
While waiting for using the specifications: XDG on linux and programdata on Windows, a step by step workaround:

```powershell
#example of modules
$paquets="Plaster,Pester,InvokeBuild,PowerShellGet,ModuleBuilder"
#setup
$paquets -split ',' |%{Install-Module $_ -Scope AllUsers -verbose  -force}
#option1: first line of PSModulePath
$paquets -split ',' |%{Find-Package  $_ | Save-Module -Path  $([Environment]::GetEnvironmentVariable("PSModulePath") -split $([System.IO.Path]::PathSeparator))[0] -verbose}
#option2: otherwise 
$paquets -split ',' |%{Find-Package  $_ | Save-Module -Path "$env:ProgramData\Microsoft\PowerShell\modules\" -verbose}

#check
$paquets -split ',' |%{(gmo -l $_).path}
    # C:\ProgramData\Microsoft\Powershell\Modules\Plaster\1.1.4\Plaster.psd1
    # C:\ProgramData\Microsoft\Powershell\Modules\Pester\5.5.0\Pester.psd1
    # C:\ProgramData\Microsoft\Powershell\Modules\InvokeBuild\5.10.4\InvokeBuild.psd1
    # C:\ProgramData\Microsoft\Powershell\Modules\PowerShellGet\2.2.5\PowerShellGet.psd1
    # C:\ProgramData\Microsoft\Powershell\Modules\ModuleBuilder\3.0.0\ModuleBuilder.psd1
```
```powershell
#if there are recalcitrants, do this before the setup
#simple uninstall
$paquets -split ',' |%{Get-InstalledModule $_ -AllVersions |Uninstall-Module -Verbose -Force}
#force uninstall
$m='C:\Program Files (x86)\WindowsPowerShell\Modules\Pester'
takeown /F $m /A /R
icacls $m /reset
icacls $m /grant "*S-1-5-32-544:F" /inheritance:d /T
Remove-Item -Path $m -Recurse -Force -Confirm:$false
```

**Comment by JustinGrote** [comment]
@SteveL-MSFT since this didn't make 7.4, is it on deck for evaluation for 7.5? Looking forward to the roadmap :)

**Comment by danielniccoli** [comment]
Or `%LocalAppData%\PowerShell`.

**Comment by ohadschn** [comment]
> Can we include the scripts folder in this discussion too? Current behavior `<modulespath>\..\Scripts` seems like a hardcoded quick and dirty fix, an afterthought.
> 
> I think it should be treated separatly from the modules folder, at least have the ability to.
> 
> Maybe it could have it's own environmental variable, at least optionally?
> 
> * `PSScriptPath`?
> * Default to `PSModulePath\..\Scripts` if `PSScriptPath` isn't set?

Without this, Controlled Folder Access would still block the installation of all modules that make use of that folder (for example, https://github.com/PowerShell/SecretManagement).

**Comment by JustinGrote** [comment]
@bwilsonms of course, that would be just one of *many* reasons why this would be desirable.

**Comment by RobertKlohr** [comment]
It's now 2024.

@SteveL-MSFT, @JohnLudlow, @iSazonov, @rkeithhill, Et al. How do we keep this moving forward? It's been almost 3 years and while there seems to be consensus on the need, there does not seem to be any tangible progress toward fixing the problem.

This issue was tagged with milestone, 7.4 but that has passed. There were also several past comments about needing an RFC, but I never saw where one was started (What does RFC even mean in the context of a PowerShell setting?), and if it did can some point the way to where that work is being done?

I know this a considered a breaking change, but it can also alleviate a previous breaking change introduced by Controlled Folder Access (see post [here](https://github.com/PowerShell/PowerShell/issues/15552#issuecomment-1049035948 )) that many of us are having to work around on a daily basis.

We need this in 2024 and ideally in 7.5.

**Comment by JustinGrote** [comment]
As a workaround, ModuleFast installs to your LocalAppData by default and automatically sets up everything for you to make it work.

https://github.com/justingrote/modulefast

**Comment by JustinGrote** [comment]
> As it stands now I consider this a breaking experience as it is. We have our Documents folders redirected and scripts take forever to finish because of the constant searching for modules in our network folders. Other than scripts I specifically need it for I'll continue to use the older versions of Powershell that I don't see this issue with. As for keeping modules in the Documents folder...for the life of me I can't see how modules are considered Documents or how the Documents folder is an appropriate location for them. And not giving an option to change this behavior? This makes it a deal breaker in my opinion.

It comes from very old decision way back in old PowerShell before OneDrive was even a concept. Can't predict the future :)

**Comment by FaffeF** [comment]
<gripe>
I think this is some sort of Microsoft sickness. Some decision is made early on (place stuff in "system32" folder). Then things change, or this is no longer a great idea (stuff should be in "system64" logically). But since other developers didn't actually read Microsoft's best practice guide, they hardcode the value. Then Microsoft for some reason goes out of their way to retain compatibility with all Windows versions ever released, making things ever more unintuitive.
</gripe>

Just make the change when it's needed.

Maybe get some guys from the Microsoft naming department, they seem to like making changes? Though then you'll probably run in to the "path to long" problem...

**Comment by danielniccoli** [comment]
> > As it stands now I consider this a breaking experience as it is. We have our Documents folders redirected and scripts take forever to finish because of the constant searching for modules in our network folders. Other than scripts I specifically need it for I'll continue to use the older versions of Powershell that I don't see this issue with. As for keeping modules in the Documents folder...for the life of me I can't see how modules are considered Documents or how the Documents folder is an appropriate location for them. And not giving an option to change this behavior? This makes it a deal breaker in my opinion.
> 
> It comes from very old decision way back in old PowerShell before OneDrive was even a concept. Can't predict the future :)

Which is why it is so important to stick to the OS guidelines and put that stuff in LOCAL APP DATA, instead of dreaming up new locations that WILL break things when new features are introduced, like OneDrive.

**Comment by KirillOsenkov** [comment]
Is there a way for a user to opt-out of OneDrive being added to PSModulesPath every time? Can we add a setting somewhere? It should be easy to add a setting and keep the old behavior by default.

**Comment by JustinGrote** [comment]
It's not OneDrive that gets added, it's the Documents folder path that onedrive is redirecting. The "simple" solution is to not redirect your My Documents to OneDrive, and just use OneDrive separately as a separate folder, there aren't any "good" alternatives that don't result in a lot of errors and I've tried a lot (symlinking, hardlinking, regedits, excludes, you name it).

**Comment by KirillOsenkov** [comment]
Can we add an option that adds a custom user-specified directory instead of Documents?

**Comment by JustinGrote** [comment]
That already exists, unfortunately it has to be done in a `powershell.config.json` file at startup in the same directory as pwsh.exe and cannot be configured at runtime.
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-7.4#psmodulepath

I requested that path be exposable but that was closed in favor of this issue:
https://github.com/PowerShell/PowerShell/issues/18613

And PowerShellGet `Install-Module` and `Install-PSResource` are hard-coded still to install to Documents despite this setting, those have to be updated to do it as well.

My `ModuleFast` module installs to localappdata by default and updates your PowerShell path and profile accordingly, so it is an alternative option in the meantime:
https://github.com/justingrote/modulefast

**Comment by KirillOsenkov** [comment]
is there a bug to support this setting in Install-Module?

**Comment by JustinGrote** [comment]
> is there a bug to support this setting in Install-Module?

https://github.com/PowerShell/PSResourceGet/issues/1494

**Comment by fsackur** [comment]
I have successfully used symlinks to solve OneDrive for hundreds of users.

Documents was redirected to OneDrive by policy. We created a folder in the user profile, moved modules there incrementally, and symlinked modules individually into Docs/blah/Modules. Modules can be moved on the fly, even if there are locked DLLs, with `[IO.File]::Move` (or `Move-Item` in PSv7). You also need to update the `InstalledLocation` in `PSGetModuleInfo.xml` if you want `Update-Module` to work.

**Comment by JustinGrote** [comment]
@fsackur pretty sure onedrive still complains of a sync conflict when you do this and there was no way even with exclusions to suppress it, unless something's been updated since I last attempted.

**Comment by mikhey** [comment]
@rkeithhill I agree it becomes an issue for users that don't have the setting to see hidden folders.  That said the `%LOCALAPPDATA%` can be used in the File Explorer address bar and should work (in theory™).  

I'm liking the `~/AppData/Local` or `~/AppData/Roaming` folders as it may line up with the ~`/.local` folder on Linux machines essentially.  Ultimately, the sooner this can be configurable either via environment variable and/or config file would be fantastic!

**Comment by pl4nty** [comment]
OneDrive follows symlinks/junctions without errors on the latest versions (tested on MSIT Fast, Insider, and prod), but the status on files/folders is syncing even when sync has completed. I've been storing `PowerShell` and `WindowsPowerShell` on a Dev Drive using junctions for a few months.

There's a new [setting to exclude folders by name](https://learn.microsoft.com/en-us/sharepoint/use-group-policy#exclude-specific-kinds-of-folders-from-being-uploaded) and ADMX/ADMLs, but it doesn't activate even when using the example values. The file-based exclusion setting works with similar values. ProcMon shows `OneDrive.exe` loading `EnableODIgnoreFolderListFromGPO` values at startup, and I can see the codepath when decompiled, but symbols aren't available so I haven't debugged any further.

**Comment by JustinGrote** [comment]
> @rkeithhill I agree it becomes an issue for users that don't have the setting to see hidden folders. That said the `%LOCALAPPDATA%` can be used in the File Explorer address bar and should work (in theory™).
> 
> I'm liking the `~/AppData/Local` or `~/AppData/Roaming` folders as it may line up with the ~`/.local` folder on Linux machines essentially. Ultimately, the sooner this can be configurable either via environment variable and/or config file would be fantastic!

I would argue ideally users don't need to ever see this folder except for advanced troubleshooting, similar to the registry, and only interface with it using the install/update/uninstall commands indirectly.

**Comment by ALIENQuake** [comment]
> OneDrive follows symlinks/junctions without errors on the latest versions (tested on MSIT Fast, Insider, and prod), but the status on files/folders is syncing even when sync has completed. I've been storing `PowerShell` and `WindowsPowerShell` on a Dev Drive using junctions for a few months.

@pl4nty  Can you please elaborate? Which Windows/OneDrive version are you using?

**Comment by danielniccoli** [comment]
> I would argue ideally users don't need to ever see the modules folder except for advanced troubleshooting, similar to the registry, and only interface with it using the install/update/uninstall commands indirectly.

Factually incorrect. AppData is and has been the location to store application data and configuration (besides the registry) for over 20 years. It's only today that more and more inexperienced Windows developers starting abusing the profile folder to place configuration data, because they're used to macOs or Linux.

> AppData: "The directory that serves as a common repository for application-specific data for the current […] user."

![grafik](https://github.com/PowerShell/PowerShell/assets/2971735/6e940a41-a071-4652-a177-b58f293e2e20)

https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-6.0#system-environment-specialfolder-applicationdata

Furthermore:

![grafik](https://github.com/PowerShell/PowerShell/assets/2971735/57c0a247-d274-47cb-bb3d-970773109239)

https://learn.microsoft.com/en-us/dotnet/api/system.environment.specialfolder?view=net-6.0

**Comment by JustinGrote** [comment]
> Factually incorrect. AppData is and has been the location to store application data and configuration (besides the registry) for over 30 years. 

Powershell is an application. Modules are application data. Therefore that's where modules should go instead of documents. Maybe you misunderstood what I was saying?

**Comment by ALIENQuake** [comment]
@pl4nty Well, ive tried and I'm still seeing "The 'ExampleFolderJunction' in Documents is junction point ..." error which prevents me from syncing Documents folder. Same for latest insider. Thanks for help anyway, I won't nagg you about this anymore.

**Comment by QuickScoP3s** [comment]
I just came by this thread, because I'm experiencing the issue where OneDrive syncs modules, which interferes with the Install-Module / Update-Module commands.

With Microsoft now promoting a [Dev Drive](https://learn.microsoft.com/en-us/windows/dev-drive/), I'd love to store all modules there, like I moved all NuGet, npm, ... sources there as well.

**Comment by ctmal956** [comment]
> > As it stands now I consider this a breaking experience as it is. We have our Documents folders redirected and scripts take forever to finish because of the constant searching for modules in our network folders. Other than scripts I specifically need it for I'll continue to use the older versions of Powershell that I don't see this issue with. As for keeping modules in the Documents folder...for the life of me I can't see how modules are considered Documents or how the Documents folder is an appropriate location for them. And not giving an option to change this behavior? This makes it a deal breaker in my opinion.
> 
> It comes from very old decision way back in old PowerShell before OneDrive was even a concept. Can't predict the future :)

Before I start my rant....I want to call BS on this.  How is it that Powershell 5.1 doesn't have any reference to document or onedrive folders but Powershell 7+ does (event after I try to remove them).  If it wasn't for the memory limitations in 5.1 that essentially break the use the MSGraph, I wouldn't even be looking at newer versions.

**Comment by ctmal956** [comment]
> > Before I start my rant....I want to call BS on this. How is it that Powershell 5.1 doesn't have any reference to document or onedrive folders but Powershell 7+ does (event after I try to remove them). If it wasn't for the memory limitations in 5.1 that essentially break the use the MSGraph, I wouldn't even be looking at newer versions.
> 
> Not sure what you mean, there's no explicit reference in either PSResourceGet or the PowerShell engine to Onedrive paths as far as I know, everything references documents. https://github.com/search?q=repo%3APowerShell%2FPowerShell%20onedrive&type=code https://github.com/search?q=repo%3APowerShell%2FPowerShellGet%20onedrive&type=code

That is exactly what I mean.  Everything references Documents.  Which means if the documents folder is redirected, it references the redirection which means it searches the redirection (and causes everything to be extremely slow).  I don't see any references to the Documents folder in the documentation: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_psmodulepath?view=powershell-7.4 
Of course, I'm not sure exactly what its referring to when it talks about the "inherited PSModulePath".

Since this thread is over 3 years old I've given up home on a fix or a way to remove the Documents references but I did find a workaround.  I created a Profile.ps1 file that parses the PSModulePath and removes all entries with the string "Documents".  Using this I was able run a time consuming script in about 1/2 the time it used to take because it wasn't constantly checking my redirected Documents folder for modules.

**Comment by JustinGrote** [comment]
> > <img alt="" width="48" src="https://camo.githubusercontent.com/cae6a348f2a814d15cad19ca9e6619991036baf4ddf42ff2cf6ae79885db4810/68747470733a2f2f6c6561726e2e6d6963726f736f66742e636f6d2f6d656469612f6c6f676f732f6c6f676f2d706f7765727368656c6c2d736f6369616c2e706e67">**[about PSModulePath - PowerShell](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_psmodulepath?view=powershell-7.4)**The PSModulePath environment variable contains a list of folder locations that are searched to find modules and resources.

Yep that's a similar workaround I do as well, and I utilize the `ModuleFast` module installer I wrote which defaults to a non-Documents path. https://github.com/JustinGrote/ModuleFast

**Comment by paschott** [comment]
Sad to see this is still being discussed after 3 years. PS modules really don't belong in OneDrive. If you install/upgrade on one machine, the new versions will show on some other machines, but you can't "upgrade-module" on those machines. You also end up syncing _everything_ on a new machine setup with OneDrive, including all old versions that may have been synced and not removed.

There are many ideas that have been suggested that have much merit and would allow for PowerShell to still function without needing to use the OneDrive sync.

Semi-related, I was bit by this with "Offline Files" and the modules being in My Documents. But the "Offline Files" sync was absolutely horrible and I had no idea what version(s) I would be working with at any given time. That experience was so bad I moved _everything_ off of my Documents that was resulting in lag.  There, the Modules took forever to load if I was connected via VPN because I had to round-trip to the offline files server to pull the modules. That was a poor experience because the modules were placed in a location that doesn't make sense for app modules.

**Comment by jpawlowski** [comment]
Since it wasn't mentioned here yet: There an experimental feature available since PowerShell 7.4 that is related to this topic:

[`PSModuleAutoLoadSkipOfflineFiles`](https://learn.microsoft.com/en-us/powershell/scripting/learn/experimental-features?view=powershell-7.4#psmoduleautoloadskipofflinefiles)

> With this feature enabled, if a user's PSModulePath contains a folder from a cloud provider, such as OneDrive, PowerShell no longer triggers the download of all files contained within that folder. Any file marked as not downloaded are skipped. Users who use cloud providers to sync their modules between machines should mark the module folder as Pinned or the equivalent status for providers other than OneDrive. Marking the module folder as Pinned ensures that the files are always kept on disk.

I don't think it will really help, it is cosmetics, but at least something is happening that shows someone is paying attention to the topic as such.

**Comment by Jaykul** [comment]
I don't understand why people are still pinging this issue. Rob Holt gave the answer 3 years ago that this is in the configuration.

In modern PowerShell (7.4 as of this writing) you can stop PowerShell from searching your Documents folder for modules by creating or modifying the `powershell.config.json` in your Documents\PowerShell folder to add a PSModulePath value like this:

```json
{
  "PSModulePath":"%LOCALAPPDATA%/powershell/Modules", 
}
```

Now PowerShell will use that instead of your Documents\PowerShell\Modules location.

I feel like I need to add: please do not forget the original reason these modules were in your Documents is because _you were expected to read them_. Even now that Microsoft is hosting the PowerShell Gallery, there is no real review or audit of the PowerShell modules in that gallery. There is not (yet) a public source of safe first-party modules like the MCR is for Microsoft container images.

**Comment by JustinGrote** [comment]
> I don't understand why people are still pinging this issue. Rob Holt gave the answer 3 years ago that this is in the configuration, and that works, right?

The big problem is that this requires a pwsh restart, and it also includes access to where pwsh.exe lives. If this is in Program Files, then you need admin rights to do this. So, not userspace enough.

**Comment by Jaykul** [comment]
No....
The path people want to change isn't the ALL USERS path (which would require editing the config file next to pwsh.exe) it's the CURRENT USER path, which is controlled by the config file next to your personal profile.ps1

> > I don't understand why people are still pinging this issue. Rob Holt gave the answer 3 years ago that this is in the configuration, and that works, right?
> 
> The big problem is that this requires a pwsh restart to implement, and it also requires access to the same directory where pwsh.exe lives. If this is in Program Files, then you need admin rights to do this. So, not userspace enough.
> 
> Fine if it's your desktop and you own it, but in VDI/Server/etc. scenarios, it's not great.

**Comment by JustinGrote** [comment]
@jaykul my apologies, I was not aware there was a per-user config.json that lives in your `$PROFILE.CurrentUserCurrentHost` directory.

In that regard, yeah it seems like publishing a simple module command like  `Set-ModulePath` or something to simplify the creation/editing of the json seems like the simplest path to resolve this, and then agreed it's just a matter of getting PSResourceGet/etc. to respect that path once set so you don't have to use `Save-Module` constantly

**Comment by cimares** [comment]
Just to add to one of the issues that OneDrive introduces when the Documents folder is included in the sync, and that is retention policies. My organisation defines a OneDrive retention policy of 5 years since Last modified. This means that large chunks of the PS Module path get deleted at odd times, resulting in incomplete module packages which have to be recovered from the OneDrive Recycle bin.

Moving user level modules out of Documents is a critical need when OneDrive retention is in the mix.

**Comment by JustinGrote** [comment]
I made a function that uses some private methods on the config and merges them with the API from #19422 to demonstrate what I think the actual public API should look like/behave as so that tools such as `PowerShellGet` and `ModuleFast` can all use the same relevant implicit installed modules path if not specified otherwise.

```powershell

#Fetches the module path for the current user or all users
function Get-PSModulePath ([Switch]$AllUsers) {
	$scopeType = [Management.Automation.Configuration.ConfigScope]
	$pscType = $scopeType.
	Assembly.
	GetType('System.Management.Automation.Configuration.PowerShellConfig')

	$pscInstance = $pscType.
	GetField('Instance', [Reflection.BindingFlags]'Static,NonPublic').
	GetValue($null)

	$getModulePathMethod = $pscType.GetMethod('GetModulePath', [Reflection.BindingFlags]'Instance,NonPublic')

	if ($AllUsers) {
		$getModulePathMethod.Invoke($pscInstance, $scopeType::AllUsers) ?? [Management.Automation.ModuleIntrinsics]::GetPSModulePath('BuiltIn')
	} else {
		$getModulePathMethod.Invoke($pscInstance, $scopeType::CurrentUser) ?? [Management.Automation.ModuleIntrinsics]::GetPSModulePath('User') 
	}
}
```

**Comment by Jaykul** [comment]
Honestly, I expected `[Management.Automation.ModuleIntrinsics]::GetPSModulePath` to return the path _after_ accounting for configuration. I guess I should file a bug for that...

I can't imagine any scenario where it's useful for this method to return a path that's **not** actually in the PSModulePath!

I mean, given that my user powershell.config specifies a %LocalAppData% path, the Documents based path **will not** be in PowerShell's PSModulePath ...

![image](https://github.com/user-attachments/assets/57a78934-238c-4a37-b65f-eb9a784ef20a)

**Comment by JustinGrote** [comment]
I have created a Module to simplify the editing of the powershell.config.json file as well as including my function for getting the current "primary" PSModulePath
https://github.com/JustinGrote/ModulePath

**Comment by rohancragg** [comment]
> Just to add to one of the issues that OneDrive introduces when the Documents folder is included in the sync, and that is retention policies. My organisation defines a OneDrive retention policy of 5 years since Last modified. This means that large chunks of the PS Module path get deleted at odd times, resulting in incomplete module packages which have to be recovered from the OneDrive Recycle bin.
> 
> Moving user level modules out of Documents is a critical need when OneDrive retention is in the mix.

My organisational policies prevent me from excluding any paths from OneDrive sync those excluding me from one of the possible work-arounds. I really need the modules to be stored somewhere other than OneDrive for al the reasons discussed here.

I blogged about this in 2021 and frustrating that it's still an issue. Especially when switching between different machines.
https://rohancragg.co.uk/misc/ps-module-paths/

**Comment by JustinGrote** [comment]
`My organisational policies prevent me from excluding any paths from OneDrive sync thus excluding me from one of the possible work-arounds.`

Why would it exclude you from the workaround of placing it at a separate path and adjusting your profile to include that as a PSModulePath folder? This doesn't fix PSGet/PSGetv3 and you need to use `Save-Module` instead of `Install-Module` but otherwise that should still be a valid workaround until a resolution moves forward.

If you can use PowerShell 7 this is also already fixable, and I made a module for it called `ModulePath`, again still need to use `ModuleFast` or `Save-PSModule/Resource` with an alternate path.

**Comment by cimares** [comment]
> If you can use PowerShell 7 this is also already fixable, and I made a module for it called `ModulePath`, again still need to use `ModuleFast` or `Save-PSModule/Resource` with an alternate path.

Just to provide some feedback on this @JustinGrote , the ModulePath approach made this nice and simple to explain to end users on how to update PowerShell 7 to get modules out of OneDrive, so thanks for the effort on releasing that. Now I just need to resolve PS5.1 modules  8-)

**Comment by wUEZRs** [comment]
This would be a great addition, or OneDrive should just exclude it by default. There's zero reason to cloud sync local modules. I find it interesting that PS Modules aren't in a native Windows folder by default

**Comment by JeffMill** [comment]
> 4 years and 1 month later. And still no working solution to allow end user control over the modulepath. Hope you all got let go during the recent layoffs.

@MattRose-CC  Shame on you. Grow up.

**Comment by MattRose-CC** [comment]
FOUR YEARS. Shame on Microsoft. Shame on excuse makers such as yourself. basic functionality of controlling environmental variables. Inexusable. Shame on the listless lack of direction and middle management that couldnt effectuate such a basic change in over 4 years.

**Comment by JustinGrote** [comment]
@MattRose-CC you must be fun at parties.

**Comment by JosephColvin** [comment]
Food for thought:
Be the hero, not a zero.
Be the change you want to see.

How to do it:
Take that passion and try to figure out how you can make the system work the way you want it to. Don't let that passion and energy do nothing, use it to make the change you want to see! Dig into the code, figure out where the changes need to be. Maybe you can do it without even touching the code base? Maybe via a configuration or a file monitor and hard link? Then share your solution for all to see and use. You can be the hero!

**Comment by jeroenlandheer** [comment]
Let's not derail this conversation and keep things constructive. I agree this takes a long time, but if you look at how many people are involved here and putting their hard work in making this happen, it is clearly evident that a seemingly simple change might have a lot of consequences. (I also asked about why this takes long, but compatibility is a big thing and that's a really good thing.) 

Basically there are 2 things going on here:
1. Where is PowerShell looking for modules?
2. Where do your modules get installed?

### Looking for modules
As @Jaykul  and @JustinGrote have pointed out, there are now new ways around this, see their comments [here](https://github.com/PowerShell/PowerShell/issues/15552#issuecomment-2297887053) and [here](https://github.com/PowerShell/PowerShell/issues/15552#issuecomment-2327851719). It is not 100% ideal as defaults for new installations aren't changed AFAIK, but at least there are supported ways now to change this behaviour. 

### Where your modules get placed
To prevent your modules land in your OneDrive folder in the first place, another issue is being worked on. This is [issue 1494](https://github.com/PowerShell/PSResourceGet/issues/1494), which involves changes to the module installer. 

If you're looking for the older commands like `Install-Module`... this is PowerShellGet, which is provided for compatibility.

> This documentation covers version 3.0.22-beta22 of the PowerShellGet module. This module is provided for compatibility with PowerShellGet v2.2.x. The cmdlets in this version of the module are proxy cmdlets that call the equivalent cmdlets in the Microsoft.PowerShell.PSResourceGet module.

Source: [Documentation](https://learn.microsoft.com/en-us/powershell/module/powershellget/?view=powershellget-3.x)

**Comment by kilasuit** [comment]
Being discussed in https://github.com/PowerShell/PowerShell-RFC/pull/388
So this issue I think should really be closed & locked at this point
